### PR TITLE
[feature](Nereids) add graph simplifier

### LINKED ISSUE #13903: [Feature] Add graph simplifier
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Now that the hypergraph is ready, we need to further simplify the graph. Therefore, a graph simplifier is needed that support to decide the order of join according the cost.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by keanji-x**
# Proposed changes

Issue Number: close #13903 

## Problem summary

Add a graph simplifier that can simplfy the hypergraph according cost.
Now we only support inner join, the very simple cost function.
And the circle detector has not been added

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.2 seconds
 load time: 571 seconds
 storage size: 17154655393 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221102115816_clickbench_pr_38296.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/GraphSimplifier.java [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+
+        List<PhysicalProperties> inputs = new ArrayList<>();
+        inputs.add(PhysicalProperties.ANY);
+        inputs.add(PhysicalProperties.ANY);
+        groupExpression.updateLowestCostTable(PhysicalProperties.ANY, inputs, cost);
+
+        return (LogicalJoin) newJoin.withGroupExpression(Optional.of(groupExpression));
+    }
+
+    private void proposeJoin(int edgeIndex, Plan plan1, Plan plan2) {
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/GraphSimplifier.java:373]: setJoin()

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/GraphSimplifier.java [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+     * is under the limit.
+     * In hyper graph, counting the csg-cmp pair is more expensive than the apply simplification step, therefore
+     * we use binary search to find the least step we should apply.
+     *
+     * @param limit the limit number of the csg-cmp pair
+     */
+    public boolean simplifyGraph(int limit) {
+        // Now we only support simplify graph until the hyperGraph only contains one plan
+        Preconditions.checkArgument(limit == 1);
+        while (applySimplificationStep()) {}
```
[the remark this answers is not in the document]
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/GraphSimplifier.java:138]: `simplifyGraph` is used for binary search, although we has not implenmented yet. It's little different from `applySimplificationStep`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/Edge.java [hunk 3]
```diff
@@ -22,20 +22,19 @@
 import java.util.BitSet;
 
 class Edge {
-    final int index;
+    int index;
```
[the remark this answers is not in the document]
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/Edge.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/GraphSimplifier.java [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+    GraphSimplifier(HyperGraph graph) {
+        this.graph = graph;
+        edgeSize = graph.edges.size();
+        for (int i = 0; i < edgeSize; i++) {
+            BestSimplification bestSimplification = new BestSimplification();
+            Plan leftPlan = graph.getPlan(graph.getEdge(i).getLeft());
+            Plan rightPlan = graph.getPlan(graph.getEdge(i).getRight());
+            joinPlans.add(Pair.of(leftPlan, rightPlan));
+            simplifications.add(bestSimplification);
+        }
```
[the remark this answers is not in the document]
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/GraphSimplifier.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/GraphSimplifier.java [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+            bestSimplification.bestNeighbor = -1;
+            Optional<SimplificationStep> optionalStep = makeSimplificationStep(edgeIndex1, edgeIndex2);
+            if (!optionalStep.isPresent()) {
+                continue;
+            }
+            proposeSimplificationStep(optionalStep.get(), bestSimplification, edgeIndex2);
+        }
+    }
+
+    private boolean proposeSimplificationStep(SimplificationStep step, BestSimplification bestSimplification,
```
[the remark this answers is not in the document]
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/GraphSimplifier.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/GraphSimplifier.java (hunk 1 above)
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/joinreorder/hypergraph/GraphSimplifier.java]: done

**Pull request closed without merging**